numba/core/caching.py

```diff
@@ -308,17 +310,63 @@ def from_function(cls, py_func, py_file):
         return self
 
 
+class _ZipCacheLocator(_SourceFileBackedLocatorMixin, _CacheLocator):
+    """
+    A locator for functions backed by Python modules within a zip archive.
+    """
+
+    def __init__(self, py_func, py_file):
+        self._py_file = py_file
+        self._lineno = py_func.__code__.co_firstlineno
+        self._zip_path, self._internal_path = self._split_zip_path(py_file)
+        # We use AppDirs at the moment. A more advanced version of this could also allow
+        # a provided `cache_dir`, though that starts to create (cache location x source
+        # type) number of cache classes.
+        appdirs = AppDirs(appname="numba", appauthor=False)
+        cache_dir = appdirs.user_cache_dir
+        cache_subpath = self.get_suitable_cache_subpath(py_file)
+        self._cache_path = os.path.join(cache_dir, cache_subpath)
+
+    @staticmethod
+    def _split_zip_path(py_file):
+        path = Path(py_file)
+        for i, part in enumerate(path.parts):
+            if part.endswith(".zip"):
+                zip_path = str(Path(*path.parts[: i + 1]))
+                internal_path = str(Path(*path.parts[i + 1 :]))
+                return zip_path, internal_path
+        return None, None
+
+    def get_cache_path(self):
+        return self._cache_path
+
+    def get_source_stamp(self):
+        if self._zip_path:
+            st = os.stat(self._zip_path)
+            return st.st_mtime, st.st_size
```

Comment:
This `if` statement will be always true, no? `from_function` will guarantee the `_ZipCacheLocator` will only be created when a zip file is given. On the ctor, `self._zip_path` will never be `None` because there will be a zip subpath.

Reply:
Yes, agree, thanks for reviewing so thoroughly. The code is now simplified.